Automatically generate properties for packet fields

TSRBerry opened the pull request:
This PR adds a source generator to generate property implementations for packet fields.

C# 13 (.NET 9.0) adds support for partial properties and source generators are able to provide implementations for them.
This generator looks for the new `PacketField` attribute which can be applied to properties and generates the appropriate implementation based on the provided arguments and the property type.

It handles every packet field that's currently implemented by RyuSOCKS.

The generator currently supports the following features:

- Constant and dynamic offsets for packet fields
- Constant and dynamic lengths for packet fields
- Little and big endian support for packet fields
- Generation for the following property types:
  - simple numeric types (`byte`, `sbyte`, `ushort`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`)
  - enums
  - strings
  - arrays of enums and simple numeric types
  - System.Net.IPAddress 
- Min/Max length checks for arrays and strings
- Invocation of custom validation methods before executing the setter/getter functions

---

There are a few things that I want to do before this gets merged:

- [ ] Add tests
- [ ] Check other projects like https://github.com/dotpcap/packetnet and figure out if it makes sense to turn this generator into a standalone NuGet package
- [x] Fix all the FIXME comments
- [ ] Create issues for all TODO comments
- [ ] Consider implementing `AddVerificationMethodIfNecessary()` in a different way to make it less awkward to use
- [ ] Restructure the helper properties and methods of the RyuSOCKS packets

TSRBerry committed:
Add source generator for Packet-type classes

TSRBerry committed:
Replace packet properties with PacketFieldAttribute where possible

TSRBerry committed:
Move builder classes to a new subfolder

TSRBerry committed:
Split PacketGenerator into multiple files

TSRBerry committed:
Fix keyword order for Accessibility.ProtectedAndInternal

TSRBerry committed:
Restructure AccessorGenerator

Hopefully it's easier to see
how the different property types are handled now.

TSRBerry committed:
Add indexer for Packet and remove PacketBytesFieldName constant

TSRBerry committed:
Use nameof() for type strings

TSRBerry committed:
PacketGenerator: Add support for strings

TSRBerry committed:
Apply PacketField to string properties

TSRBerry committed:
PacketGenerator: Add support for simple array types

TSRBerry committed:
Replace RegisterSourceOutput with RegisterImplementationSourceOutput

TSRBerry committed:
PacketGenerator: Add AssumeGeneratedEnumType to field attribute

TSRBerry committed:
Apply PacketField to generated and array properties

TSRBerry committed:
PacketGenerator: Remove TODO comment about grouping generated properties

TSRBerry committed:
PacketGenerator: Add ValidationMethod to packet field attribute

TSRBerry committed:
Apply PacketField to properties with extra logic

TSRBerry committed:
PacketGenerator: Add special handling for IPAddress properties

TSRBerry committed:
Apply PacketField to IPAddress property

TSRBerry committed:
Fix whitespace issues in post initialization sources

TSRBerry committed:
Remove global alias from fully qualified names

TSRBerry committed:
Fix code generation for property access modifiers

TSRBerry committed:
Create test project for RyuSocks.Generator

TSRBerry committed:
PacketGenerator: Add tests for simple numeric property types

TSRBerry committed:
Add source generator testing package to Directory.Packages.props

TSRBerry committed:
PacketGenerator: Only assign to LengthMember if allowed

TSRBerry committed:
PacketGenerator: Generate correct accessor modifiers

TSRBerry committed:
PacketGenerator: Add simple string tests

AnonymusNW commented on RyuSocks.Generator/PacketGenerator.cs:
```diff
@@ -0,0 +1,516 @@
+
+            foreach (var accessor in propertyDeclarationSyntax.AccessorList!.Accessors)
+            {
+                if (accessor.IsKind(accessorKind))
```
Invert if statement to keep the style of the previous function

AnonymusNW commented on RyuSocks.Generator/Packet/AccessorGenerator.Array.cs:
```diff
@@ -0,0 +1,227 @@
+                source.AppendLine($"result[i] = {maybeCast}this[{packetField.GetOffset()} + i];");
+                source.LeaveScope();
+                source.AppendLine();
+                source.AppendLine("return result;");
```
Add newlines to improve discernability between normal added lines and scope
```suggestion
                source.AppendLine($"{fieldTypeName}[] result = new {fieldTypeName}[{packetField.GetLength()}];");
                source.AppendLine();
                
                source.EnterScope($"for (int i = 0; i < {packetField.GetLength()}; i++)");
                source.AppendLine($"result[i] = {maybeCast}this[{packetField.GetOffset()} + i];");
                source.LeaveScope();
                
                source.AppendLine();
                source.AppendLine("return result;");
```

AnonymusNW commented on RyuSocks.Generator/PacketGenerator.cs:
```diff
@@ -0,0 +1,516 @@
+                : string.Empty;
+            string validationMethodName = !validationMethodArg.IsNull
+                ? (string)validationMethodArg.Value!
+                : string.Empty;
```
Split into two blocks and add comments for both blocks to improve readability
```suggestion
            // TypedConstants to acquire information used in the initialization of latter variables
            TypedConstant lengthArg = attributeData.NamedArguments.SingleOrDefault(kvp => kvp.Key == "Length").Value;
            TypedConstant lengthMemberArg = attributeData.NamedArguments.SingleOrDefault(kvp => kvp.Key == "LengthMember").Value;
            TypedConstant isBigEndianArg = attributeData.NamedArguments.SingleOrDefault(kvp => kvp.Key == "IsBigEndian").Value;
            TypedConstant minLengthArg = attributeData.NamedArguments.SingleOrDefault(kvp => kvp.Key == "MinLength").Value;
            TypedConstant maxLengthArg = attributeData.NamedArguments.SingleOrDefault(kvp => kvp.Key == "MaxLength").Value;
            TypedConstant assumeGeneratedEnumTypeArg = attributeData.NamedArguments.SingleOrDefault(kvp => kvp.Key == "AssumeGeneratedEnumType").Value;
            TypedConstant validationMethodArg = attributeData.NamedArguments.SingleOrDefault(kvp => kvp.Key == "ValidationMethod").Value;
            
            // Latter variables initialized via TypedConstants
            int length = !lengthArg.IsNull ? (int)lengthArg.Value! : -1;
            string lengthMember = !lengthMemberArg.IsNull ? (string)lengthMemberArg.Value! : string.Empty;
            ActualType lengthMemberType = ActualType.NamedType;
            bool isBigEndian = !isBigEndianArg.IsNull && (bool)isBigEndianArg.Value!;
            int minLength = !minLengthArg.IsNull ? (int)minLengthArg.Value! : -1;
            int maxLength = !maxLengthArg.IsNull ? (int)maxLengthArg.Value! : -1;
            string assumeGeneratedEnumType = !assumeGeneratedEnumTypeArg.IsNull
                ? (string)assumeGeneratedEnumTypeArg.Value!
                : string.Empty;
            string validationMethodName = !validationMethodArg.IsNull
                ? (string)validationMethodArg.Value!
                : string.Empty;
```

TSRBerry replied:
I addressed this by moving the code into a new class and adding some comments.
I hope this is easier to read now.

AnonymusNW commented on RyuSocks.Generator/Packet/ModelExtensions.cs:
```diff
@@ -0,0 +1,54 @@
+ * along with this program. If not, see <https://www.gnu.org/licenses/>.
+ */
+
+using Microsoft.CodeAnalysis;
```
Not required by the code
```suggestion
```

TSRBerry committed:
Add missing copyright header

TSRBerry committed:
Remove unused import

TSRBerry committed:
Refactor Exception generation for better compatability

TSRBerry committed:
Add test for simple string with writable LengthMember

TSRBerry committed:
Remove PostInitGeneratedSources from SourceGeneratorVerifier

TSRBerry committed:
Add tests for simple properties with OffsetMember

TSRBerry committed:
Replace BitConverter with BinaryPrimitives

The generated code now invokes the correct endian-specific
BinaryPrimitives Read/Write methods.

TSRBerry committed:
PacketGenerator: Add tests for namespaces

TSRBerry committed:
Apply dotnet format changes

AnonymusNW commented on RyuSocks.Generator.Test/SourceGeneratorVerifier.cs:
```diff
@@ -0,0 +1,49 @@
+                var commandLineArguments = CSharpCommandLineParser.Default.Parse(args, baseDirectory: Environment.CurrentDirectory, sdkDirectory: Environment.CurrentDirectory);
+                var nullableWarnings = commandLineArguments.CompilationOptions.SpecificDiagnosticOptions;
+
+                return nullableWarnings;
```
why not just return the value?
```suggestion
                return commandLineArguments.CompilationOptions.SpecificDiagnosticOptions;
```

AnonymusNW commented on RyuSocks.Generator/Packet/AccessorGenerator.Array.cs:
```diff
@@ -0,0 +1,192 @@
+                source.AppendLine("return result;");
+            }
+            else
+            {
+                source.EnterScope($"for (int i = 0; i < {packetField.GetLength()}; i++)");
+                source.AppendLine($"this[{packetField.GetOffset()} + i] = (byte)value[i];");
+                source.LeaveScope();
+            }
+
+            return source.GetLines();
+        }
+
+        private static string[] GenerateArrayIntegralAccessor(PacketFieldModel packetField, TypeConverterModel converter, bool isGetter)
+        {
+            BlockBuilder source = new();
+
+            if (isGetter)
+            {
+                string fieldTypeName = packetField.FieldType.Name.Extract(0, -2);
+                string maybeCast = packetField.FieldType.IsEnum ? $"({fieldTypeName})" : string.Empty;
+
+                source.AppendLine($"{fieldTypeName}[] result = new {fieldTypeName}[{packetField.GetLength()}];");
+                source.AppendLine();
+                source.EnterScope($"for (int i = 0; i < {packetField.GetLength()}; i++)");
+                source.AppendLine($"result[i] = {maybeCast}{converter.GetReaderName(packetField.IsBigEndian)}(this.AsSpan({packetField.GetOffset()} + (i * {converter.Length}), {converter.Length}));");
+                source.LeaveScope();
+                source.AppendLine();
+                source.AppendLine("return result;");
+            }
+            else
+            {
+                string valueParameter = packetField.FieldType.IsEnum ? $"({packetField.FieldType.ActualType.ToTypeString()})value" : "value";
+
+                source.EnterScope($"for (int i = 0; i < {packetField.GetLength()}; i++)");
+                source.AppendLine($"{converter.GetWriterName(packetField.IsBigEndian)}(this.AsSpan({packetField.GetOffset()} + (i * {converter.Length}), {converter.Length}), {valueParameter}[i]);");
+                source.LeaveScope();
+            }
+
+            return source.GetLines();
+        }
```
In all of these 3 lines the Block Builder gets set up the same way around the one main changing line, is it possible to make this a function?

TSRBerry replied:
I extracted all the common code to `GenerateSimpleArrayAccessor()` and simplified the other methods instead.

AnonymusNW reviewed (commented):
Made some comments about possible improvements.
Also why are there multiple classes just called "TestPacket" or similiar?
In e.g.: RyuSocks.Generator.Test/data/PacketGenerator/Namespaces/TestPacket.cs

TSRBerry commented:
> Made some comments about possible improvements. Also why are there multiple classes just called "TestPacket" or similiar? In e.g.: RyuSocks.Generator.Test/data/PacketGenerator/Namespaces/TestPacket.cs

Thank you very much! I'll address your comments shortly.

All the files in `RyuSocks.Generator.Test/data` are used by `PacketGeneratorTests.GeneratedSources_AsExpected()`. It reads the source file, runs the generator and compares the results to the expected generated files (and also checks for compilation errors).
I recently learned that this seems to be called snapshot testing?

TSRBerry committed:
Invert if-statement to reduce nesting

TSRBerry committed:
Return nullableWarnings directly

TSRBerry committed:
Move attribute parsing to a new class

TSRBerry committed:
Extract common code from AccessorGenerator.Array methods

TSRBerry committed:
Change return type to string for some AccessorGenerator.Simple methods

AnonymusNW reviewed (approved):
lgtm

TSRBerry committed:
Remove static default class in PacketFieldAttributeData

TSRBerry committed:
Add note to keep default values in sync

TSRBerry committed:
Add enum for string encodings to PacketFieldAttribute

TSRBerry committed:
Generate encoding specific string accessors

TSRBerry committed:
Add tests for different string encodings

TSRBerry committed:
Remove enum string generation code

The affected code is unreachable.

TSRBerry commented:
With the latest commits, all `FIXME` comments have been fixed!

We now support all string encodings which are exposed by [System.Text.Encoding](https://learn.microsoft.com/en-us/dotnet/api/system.text.encoding?view=netstandard-2.1). 
.NET Core also exposes `Latin1` encoding, which is not supported, since we currently aren't able to test it properly.

The test coverage for the new encoding code is ~99%, the uncovered code is about exceptions, which I plan to add tests for soon.
And with that we also achieved ~99% test coverage of `AccessorGenerator.Simple.cs`!

AnonymusNW commented on RyuSocks.Generator/Packet/ExceptionHelper.cs:
```diff
@@ -0,0 +1,57 @@
+    {
+        public static class ArgumentOutOfRange
+        {
+            public static void GenerateThrowIfNotEqual(BlockBuilder source, string value, string other, string overrideValueName = "", string overrideValueText = "")
+            {
+                string nameofText = overrideValueName.Length > 0 ? overrideValueName : value;
+                string valueText = overrideValueText.Length > 0 ? overrideValueText : $"{{nameof({value})}}";
+                string message = $"$\"{valueText} must be equal to: {{{other}}}\"";
+                source.EnterScope($"if ({value} != {other})");
+                source.AppendLine($"throw new ArgumentOutOfRangeException(nameof({nameofText}), {value}, {message});");
+                source.LeaveScope();
+            }
+
+            public static void GenerateThrowIfLessThan(BlockBuilder source, string value, string other, string overrideValueName = "", string overrideValueText = "")
+            {
+                string nameofText = overrideValueName.Length > 0 ? overrideValueName : value;
+                string valueText = overrideValueText.Length > 0 ? overrideValueText : $"{{nameof({value})}}";
+                string message = $"$\"{valueText} must be larger or equal to: {{{other}}}\"";
+                source.EnterScope($"if ({value} < {other})");
+                source.AppendLine($"throw new ArgumentOutOfRangeException(nameof({nameofText}), {value}, {message});");
+                source.LeaveScope();
+            }
+
+            public static void GenerateThrowIfGreaterThan(BlockBuilder source, string value, string other, string overrideValueName = "", string overrideValueText = "")
+            {
+                string nameofText = overrideValueName.Length > 0 ? overrideValueName : value;
+                string valueText = overrideValueText.Length > 0 ? overrideValueText : $"{{nameof({value})}}";
+                string message = $"$\"{valueText} must be smaller or equal to: {{{other}}}\"";
+                source.EnterScope($"if ({value} > {other})");
+                source.AppendLine($"throw new ArgumentOutOfRangeException(nameof({nameofText}), {value}, {message});");
+                source.LeaveScope();
+            }
```
this is the same function thrice (safe for one word changed in string message)
it should be possible to make this one function and change the text with a switch statement
if it is important that the three functions stay, just make them call the one additional function with the corresponding case set:
```
public static void GenerateThrowIfLessThan(BlockBuilder source, string value, string other, string overrideValueName = "", string overrideValueText = "", int case)
string nameofText = overrideValueName.Length > 0 ? overrideValueName : value;
                string valueText = overrideValueText.Length > 0 ? overrideValueText : $"{{nameof({value})}}";
               switch(case)
               {case 1:
               ...}
                source.EnterScope($"if ({value} < {other})");
                source.AppendLine($"throw new ArgumentOutOfRangeException(nameof({nameofText}), {value}, {message});");
                source.LeaveScope();
```

TSRBerry replied:
That's reasonable, I'll do that thanks!

AnonymusNW commented on RyuSocks.Generator/Packet/StringEncodingExtensions.cs:
```diff
@@ -0,0 +1,77 @@
+        {
+            if (isBigEndian)
+            {
+                switch (stringEncoding)
+                {
+                    case StringEncoding.Unicode:
+                        return [$"Encoding.BigEndianUnicode.GetBytes({sourceArgText}, {destinationArgText});"];
+                    case StringEncoding.UTF32:
+                        BlockBuilder source = new();
+
+                        // Write bytes in little endian to temp array
+                        source.AppendLine($"byte[] byteArray = Encoding.UTF32.GetBytes({sourceArgText});");
+                        source.AppendLine("Array.Reverse(byteArray);");
+                        // Write bytes with correct endianness to the packet
+                        source.AppendLine($"byteArray.CopyTo({destinationArgText});");
+
+                        return source.GetLines();
+                }
```
Is this switch not covering all cases planned?

TSRBerry replied:
Yep, there are only two special cases that need to be treated differently.
The rest is covered by the return statement below.

AnonymusNW reviewed (approved):
two remarks, else lgtm

TSRBerry committed:
ExceptionHelper: Move common code to a new method

TSRBerry committed:
Add ExcludeFromCodeCoverage attribute to relevant classes

TSRBerry committed:
Make TypeConverterModel readonly

TSRBerry committed:
Parse both "variants" of actual type names

TSRBerry committed:
Add expected diagnostics to VerifyGeneratedSources

TSRBerry committed:
Add exception test for bad attribute constructor args

TSRBerry committed:
Add test for generated enum types

TSRBerry committed:
Make builder classes public and rename IndentCount to IndentLevel

CurrentIndentCount => CurrentIndentLevel
Remove ExcludeCodeCoverage attribute

TSRBerry committed:
Add unit tests for builder classes

TSRBerry committed:
Abstract ToString() and add missing newline to BlockBuilder.ToString()

TSRBerry committed:
Add justifications for ExcludeFromCodeCoverage usage

TSRBerry committed:
Make SourceGenerator test class private

TSRBerry committed:
Move simple array accessors to a new file

TSRBerry committed:
Generate accessors for structs

TSRBerry committed:
Generate accessors for string arrays
